### PR TITLE
Implement multi-process pipelines

### PR DESCRIPTION
**Pull request by free-variation**
A PR that (unfortunately) touches many files in order for a pipeline to chain the execution of multiple Process objects.  

A `Pipeline` has a list of `Process` objects.  A concrete `Process` class provides an `algorithm` method that takes a `Doc` as input and returns a `Doc`,  

The NLP object, which acts as the entry point to the CLTK v1's API, has a pipeline.  When its `analyze` method is invoked on a text string, it:

- Creates a fresh `Doc` object, setting the `raw` attribute to the input text;
- Loops over the `Process` objects in NLP's `Pipeline`, and for each one:
  - creates an instance of the `Process`, setting its `input_doc` attribute to the current `Doc`
  - invokes the `Process.run` method, which in turn fires the `Process.algorithm` method and sets the `output_doc` attribute to the result.
  - updates the current doc to the `Process`' `output_doc`

Thus the currency that passes between `Process` elements is the `Doc`, whose attributes the processes are free to alter.

### CODE REVIEW

**Comment by free-variation** [comment]
Example:

   ```python
 >>> from cltkv1 import NLP
    >>> nlp = NLP(language='got')
    >>> from cltkv1.dependency.tree import TreeBuilderProcess
    >>> nlp.pipeline.add_process(TreeBuilderProcess)
    >>> from cltkv1.utils.example_texts import EXAMPLE_TEXTS

    >>> doc = nlp.analyze(text=EXAMPLE_TEXTS["got"])
    >>> doc.trees[0].print_tree()

root | liuhtjai_2 [lemma=liuhtjan,pos=V-,upos=VERB,xpos=V-,Mood=Opt,Number=Sing,Person=3,Tense=Pres,VerbForm=Fin,Voice=Act]
    └─ advmod | swa_1 [lemma=swa,pos=Df,upos=ADV,xpos=Df]
    └─ nsubj | liuhaþ_3 [lemma=liuhaþ,pos=Nb,upos=NOUN,xpos=Nb,Case=Nom,Gender=Neut,Number=Sing]
        └─ nmod | izwar_4 [lemma=izwar,pos=Ps,upos=ADJ,xpos=Ps,Case=Nom,Gender=Neut,Number=Sing,Person=2,Poss=Yes]
    └─ obl | andwairþja_6 [lemma=andwairþi,pos=Nb,upos=NOUN,xpos=Nb,Case=Dat,Gender=Neut,Number=Sing]
        └─ case | in_5 [lemma=in,pos=R-,upos=ADP,xpos=R-]
        └─ nmod | manne,_7 [lemma=manna#2,pos=Nb,upos=NOUN,xpos=Nb,Case=Gen,Gender=Masc,Number=Plur]
    └─ ccomp | gasaiƕaina_9 [lemma=ga-saiƕan,pos=V-,upos=VERB,xpos=V-,Mood=Opt,Number=Plur,Person=3,Tense=Pres,VerbForm=Fin,Voice=Act]
        └─ mark | ei_8 [lemma=ei,pos=G-,upos=SCONJ,xpos=G-]
        └─ iobj | izwara_10 [lemma=jūs,pos=Pp,upos=PRON,xpos=Pp,Case=Gen,Gender=Masc,Number=Plur,Person=2,PronType=Prs]
        └─ obj | waurstwa_12 [lemma=waurstw,pos=Nb,upos=NOUN,xpos=Nb,Case=Acc,Gender=Neut,Number=Plur]
            └─ amod | goda_11 [lemma=goþs,pos=A-,upos=ADJ,xpos=A-,Case=Nom,Degree=Pos,Gender=Neut,Number=Plur,Strength=Strong]
        └─ cc | jah_13 [lemma=jah,pos=C-,upos=CCONJ,xpos=C-]
        └─ conj | hauhjaina_14 [lemma=hauhjan,pos=V-,upos=VERB,xpos=V-,Mood=Opt,Number=Plur,Person=3,Tense=Pres,VerbForm=Fin,Voice=Act]
            └─ obj | attan_15 [lemma=atta,pos=Nb,upos=NOUN,xpos=Nb,Case=Acc,Gender=Masc,Number=Sing]
                └─ nmod | izwarana_16 [lemma=izwar,pos=Ps,upos=ADJ,xpos=Ps,Case=Acc,Gender=Masc,Number=Sing,Person=2,Poss=Yes]
                └─ det | þana_17 [lemma=sa,pos=Pd,upos=DET,xpos=Pd,Case=Acc,Gender=Masc,Number=Sing]
                └─ nmod | himinam._19 [lemma=himins,pos=Nb,upos=NOUN,xpos=Nb,Case=Dat,Gender=Masc,Number=Plur]
                    └─ case | in_18 [lemma=in,pos=R-,upos=ADP,xpos=R-]
>>>
```

**Comment by free-variation** [comment]
@kylepjohnson A few scattered notes:

- I removed the DefaultPipeline, since I couldn't think clearly about what a default might mean in the absence of a language.
- I removed `MultiProcess`, because I didn't understand the concept (which isn't a good reason), since multiple processes presumably exist at the `Pipeline` level, not in a single Process.   Should we add it back?
- Rather than storing a `Pipeline` class in `NLP`, I used an instance of `Pipeline`, because initialization might be expensive (e.g. loading large models), so I wanted to avoid recreating the pipeline whenever `NLP.analyze` is invoked.

**Comment by free-variation** [comment]
Also @kylepjohnson I'd like to suggest we move code of `__init__.py`.  According to online sources that file should only specify module initialization code, if any.   For the specific case of the wrappers, to my mind it's fine to put the Process into the same file as the wrapper proper.  The other case I found is in the languages module.  There we could just move the functions to a utils.py file, perhaps?

File: poetry.lock [hunk 1]
```diff
@@ -281,7 +281,7 @@ description = "IPython: Productive Interactive Computing"
 name = "ipython"
 optional = false
 python-versions = ">=3.6"
-version = "7.11.0"
+version = "7.11.1"
```
Comment (kylepjohnson) [inline poetry.lock]: A question about poetry -- right now it's set to get current declared *or newer*. Should we fix this to equal versions once in production? I imagine so.

File: src/cltkv1/dependency/tree.py [hunk 2]
```diff
@@ -236,6 +236,25 @@ def to_tree(sentence: List[Word]) -> "DependencyTree":
...
 
 
+class TreeBuilderProcess(Process):
+    """A ``Process`` that takes a doc containing sentences of CLTK words
+    and returns a dependency tree for each sentence.
+
+    >>> from cltkv1 import NLP
+    >>> nlp = NLP(language='got')
+    >>> from cltkv1.dependency.tree import TreeBuilderProcess
+    >>> nlp.pipeline.add_process(TreeBuilderProcess)
```
Comment (kylepjohnson) [inline src/cltkv1/dependency/tree.py]: Inordinately delighted by `add_process()`!

File: src/cltkv1/tokenizers/word.py [hunk 3]
```diff
@@ -45,15 +57,15 @@ class DefaultTokenizationProcess(TokenizationProcess):
 
     >>> from cltkv1.tokenizers.word import DefaultTokenizationProcess
     >>> from cltkv1.utils.example_texts import EXAMPLE_TEXTS
-    >>> tok = DefaultTokenizationProcess(data_input=EXAMPLE_TEXTS["non"][:29])
+    >>> tok = DefaultTokenizationProcess(input_doc = Doc(raw=EXAMPLE_TEXTS["non"][:29]))
     >>> tok.description
     'Whitespace tokenizer inheriting from the NLTK'
-    >>> tok.data_output
+    >>> tok.run()
+    >>> tok.output_doc.tokens
```
Comment (kylepjohnson) [inline src/cltkv1/tokenizers/word.py]: 👍

File: src/cltkv1/utils/data_types.py [hunk 4]
```diff
@@ -73,6 +72,29 @@ class Word:
...
+
+    indices_sentences: List[List[int]] = None
+    indices_tokens: List[List[int]] = None
+    language: str = None
+    words: List[Word] = None
+    pipeline: List["Process"] = None
+    raw: str = None
+
+    @property
+    def sentences(self):
```
Comment (kylepjohnson) [inline src/cltkv1/utils/data_types.py]: You have anticipated what I was imaging -- a variety of access methods/properties to pull what users will conventionally want. So `Doc.sentences` and (say) `Doc. dependency_tree` and `Doc.pos` etc. These methods would usually read from `Doc.words` and return a summarization of this info.

Are we thinking similarly here? Seem like the most comfortable interface for the average user?

File: src/cltkv1/utils/data_types.py [hunk 5]
```diff
@@ -81,58 +103,31 @@ class Process:
...
-
-    indices_sentences: List[List[int]] = None
-    indices_tokens: List[List[int]] = None
-    language: str = None
-    words: List[Word] = None
-    pipeline: List[Process] = None
-    raw: str = None
+            self.output_doc = self.algorithm(self.input_doc)
+        else:
+            raise NotImplementedError
```
Comment (kylepjohnson) [inline src/cltkv1/utils/data_types.py]: As I will note in my comment to the entire PR, I still see need of something like MultiProcess, for when a 3rd part algo does more than one kind processing (say, Lemmatization and dependency parsing).

**If we still do not want MultiProcess, then by what Process should it be wrapped?**

File: src/cltkv1/wrappers/__init__.py [hunk 6]
```diff
@@ -1,11 +1,13 @@
...
-from cltkv1.utils.data_types import Doc, MultiProcess, Word
+import copy
+
+from cltkv1.utils.data_types import Doc, Process, Word
 
 from .stanford import StanfordNLPWrapper
 
 
-class StanfordNLPProcess(MultiProcess):
+class StanfordNLPProcess(Process):
```
Comment (kylepjohnson) [inline src/cltkv1/wrappers/__init__.py]: OK, this is your answer to my above about MultiProcess. Let's go with this for now and then abstract up to MultiProcess if/when necessary.

**Comment by kylepjohnson** [comment]
The write-up above helps a ton. I can picture this before even diving into the code.

> Thus the currency that passes between Process elements is the Doc, whose attributes the processes are free to alter.

The more I think of it, this is better than having some kind of generic Process-to-Word transformer. The author of a Process should know what needs to be persisted and how to do it.

> doc.trees[0].print_tree()

The trees look fantastic.

> .add_process

So smart.

> I removed MultiProcess

I meant it for something like stanfordnlp, which does algorithms usually contained within individual Processes (tokenization, pos, depend). What kind of `Process` would you use for this?

> Rather than storing a Pipeline class in NLP, I used an instance of Pipeline, because initialization might be expensive (e.g. loading large models), so I wanted to avoid recreating the pipeline whenever NLP.analyze is invoked.

This sounds right but I'll need to look closely to make sure I understand.

> suggest we move code out of __init__.py

Not a problem. I honestly get confused about expected behavior with __init__

^^^ @free-variation I wrote this all last night, then lost connection.

Builds fine on my end.

File: poetry.lock (hunk 1 above)
Comment (free-variation) [reply poetry.lock]: I would vote yes to that.

File: src/cltkv1/utils/data_types.py (hunk 4 above)
Comment (free-variation) [reply src/cltkv1/utils/data_types.py]: Yes exactly.  `Doc.words` plus whatever extra info is needed (like the token indices for sentences) are read by useful accessors.

**Pull request merged**